### PR TITLE
[redis] fix pipeline return types

### PR DESCRIPTION
**Pull request by chdsbd**
`Pipeline` is a subclass of `Redis`, but instead of the commands returning results, they return `Pipeline`.

Previously the types said `Redis().pipeline().rpush("mylist")` returned `int`. The updated types now type the response as `Pipeline`.

With this change the following code now type checks.
```python
from redis import Redis
r = Redis()
list_length, _ = r.pipeline().rpush("mylist").ltrim("mylist", 0, 100).execute()
```

### CODE REVIEW

File: stubs/redis/redis/client.pyi [hunk 1]
```diff
@@ -684,38 +684,297 @@ class PubSub:
...
+    def __exit__(self, exc_type, exc_value, traceback) -> None: ...
+    def __del__(self) -> None: ...
+    def __len__(self) -> int: ...
     command_stack: Any
     scripts: Any
     explicit_transaction: Any
-    def reset(self): ...
+    def reset(self) -> None: ...
     def multi(self) -> None: ...
-    def execute_command(self, *args, **kwargs): ...
```
Comment (chdsbd) [inline stubs/redis/redis/client.pyi:705]: `execute_command`, along with `parse_response`, `watch`, and `unwatch` were duplicated in the methods I copied from `Redis`, so I deleted these and updated the copied ones below.

File: stubs/redis/redis/client.pyi [hunk 2]
```diff
@@ -684,38 +684,297 @@ class PubSub:
...
     def raise_first_error(self, commands, response): ...
     def annotate_exception(self, exception, number, command): ...
-    def parse_response(self, connection, command_name, **options): ...
     def load_scripts(self): ...
     def execute(self, raise_on_error: bool = ...) -> List[Any]: ...
-    def watch(self, *names: _Key) -> bool: ...
-    def unwatch(self): ...
     def script_load_for_pipeline(self, script): ...
-
-class StrictPipeline(BasePipeline, StrictRedis): ...
```
Comment (chdsbd) [inline stubs/redis/redis/client.pyi:717]: `StrictPipeline` and `BasePipeline` don't exist in the library, so I think it's safe to use use the name `Pipeline`: https://github.com/andymccurdy/redis-py/blob/1a41cfd95a53a95b078084d8627be6b6fba3bb71/redis/client.py

File: stubs/redis/redis/client.pyi [hunk 3]
```diff
@@ -684,38 +684,297 @@ class PubSub:
...
-class Pipeline(BasePipeline, Redis): ...
+    # methods from `Redis` client.
+    # arguments match `Redis`, but with some exceptions return types are
+    # `Pipeline` or `None`
```
Comment (chdsbd) [inline stubs/redis/redis/client.pyi:714]: Basically all the commands have the same arguments as `Redis`, but the return types are different.

File: stubs/redis/redis/client.pyi [hunk 4]
```diff
@@ -684,38 +684,297 @@ class PubSub:
...
+        name: _Key,
+        start: Optional[int] = ...,
+        num: Optional[int] = ...,
+        by: Optional[_Key] = ...,
+        get: Optional[Union[_Key, Sequence[_Key]]] = ...,
+        desc: bool = ...,
+        alpha: bool = ...,
+        store: Optional[_Key] = ...,
+        groups: bool = ...,
+    ) -> Pipeline: ...
```
Comment (chdsbd) [inline stubs/redis/redis/client.pyi:867]: `Redis().sort()` is overloaded to provide an accurate return type, but for `Pipeline` the return type is always `Pipeline`, so I removed the overloads.

https://github.com/python/typeshed/blob/e2fd852952ff807c60819f8840e8b23664cc7593/stubs/redis/redis/client.pyi#L495-L534

File: stubs/redis/redis/client.pyi [hunk 5]
```diff
@@ -684,38 +684,297 @@ class PubSub:
...
+    def scard(self, name: _Key) -> Pipeline: ...
+    def sdiff(self, keys, *args) -> Pipeline: ...
+    def sdiffstore(self, dest, keys, *args) -> Pipeline: ...
+    def sinter(self, keys: _Key, *args: _Key) -> Pipeline: ...
+    def sinterstore(self, dest, keys, *args) -> Pipeline: ...
+    def sismember(self, name: _Key, value: _Value) -> Pipeline: ...
+    def smembers(self, name: _Key) -> Pipeline: ...
+    def smove(self, src, dst, value) -> Pipeline: ...
+    def spop(self, name, count: Optional[int] = ...) -> Pipeline: ...
+    def srandmember(self, name: _Key, number: Optional[int] = ...) -> Pipeline: ...
```
Comment (chdsbd) [inline stubs/redis/redis/client.pyi:886]: I removed the overload here too for the same reason.

https://github.com/python/typeshed/blob/e2fd852952ff807c60819f8840e8b23664cc7593/stubs/redis/redis/client.pyi#L555-L558

File: stubs/redis/redis/client.pyi [hunk 6]
```diff
@@ -684,38 +684,297 @@ class PubSub:
...
+        num: Optional[int] = ...,
+        by: Optional[_Key] = ...,
+        get: Optional[Union[_Key, Sequence[_Key]]] = ...,
+        desc: bool = ...,
+        alpha: bool = ...,
+        store: Optional[_Key] = ...,
+        groups: bool = ...,
+    ) -> Pipeline: ...
+    def scan(self, cursor: int = ..., match: Optional[_Key] = ..., count: Optional[int] = ...) -> Pipeline: ...
+    def scan_iter(self, match: Optional[Text] = ..., count: Optional[int] = ...) -> Iterator[Any]: ...
```
Comment (chdsbd) [inline stubs/redis/redis/client.pyi:869]: The `scan_iter` and other `_iter` methods return generators that explode at runtime.

File: stubs/redis/redis/client.pyi [hunk 7]
```diff
@@ -684,38 +684,297 @@ class PubSub:
...
+    def hmget(self, name: _Key, keys: Union[_Key, Iterable[_Key]], *args: _Key) -> Pipeline: ...
+    def hvals(self, name: _Key) -> Pipeline: ...
+    def publish(self, channel: _Key, message: _Key) -> Pipeline: ...
+    def eval(self, script, numkeys, *keys_and_args) -> Pipeline: ...
+    def evalsha(self, sha, numkeys, *keys_and_args) -> Pipeline: ...
+    def script_exists(self, *args) -> Pipeline: ...
+    def script_flush(self) -> Pipeline: ...
+    def script_kill(self) -> Pipeline: ...
+    def script_load(self, script) -> Pipeline: ...
+    def register_script(self, script: Union[Text, _StrType]) -> Script: ...
```
Comment (chdsbd) [inline stubs/redis/redis/client.pyi:971]: unlike other methods, this one actually returns `Script`

File: stubs/redis/redis/client.pyi [hunk 8]
```diff
@@ -684,38 +684,297 @@ class PubSub:
...
+    def pubsub_channels(self, pattern: _Key = ...) -> Pipeline: ...
+    def pubsub_numsub(self, *args: _Key) -> Pipeline: ...
+    def pubsub_numpat(self) -> Pipeline: ...
+    def monitor(self) -> Monitor: ...
+    def cluster(self, cluster_arg: str, *args: Any) -> Pipeline: ...
+    def client(self) -> Any: ...
```
Comment (chdsbd) [inline stubs/redis/redis/client.pyi:977]: this method explodes when called

File: stubs/redis/redis/client.pyi [hunk 9]
```diff
@@ -684,38 +684,297 @@ class PubSub:
...
+    def evalsha(self, sha, numkeys, *keys_and_args) -> Pipeline: ...
+    def script_exists(self, *args) -> Pipeline: ...
+    def script_flush(self) -> Pipeline: ...
+    def script_kill(self) -> Pipeline: ...
+    def script_load(self, script) -> Pipeline: ...
+    def register_script(self, script: Union[Text, _StrType]) -> Script: ...
+    def pubsub_channels(self, pattern: _Key = ...) -> Pipeline: ...
+    def pubsub_numsub(self, *args: _Key) -> Pipeline: ...
+    def pubsub_numpat(self) -> Pipeline: ...
+    def monitor(self) -> Monitor: ...
```
Comment (chdsbd) [inline stubs/redis/redis/client.pyi:975]: This also returns `Monitor` instead of `Pipeline`

File: stubs/redis/redis/client.pyi [hunk 10]
```diff
@@ -684,38 +684,297 @@ class PubSub:
...
+        self,
+        name: _Key,
+        value: _Value,
+        ex: Union[None, int, timedelta] = ...,
+        px: Union[None, int, timedelta] = ...,
+        nx: bool = ...,
+        xx: bool = ...,
+        keepttl: bool = ...,
+    ) -> Pipeline: ...
+    def __setitem__(self, name, value) -> None: ...
```
Comment (chdsbd) [inline stubs/redis/redis/client.pyi:825]: should all the dunder methods be organized at the top?

File: stubs/redis/redis/client.pyi [hunk 11]
```diff
@@ -684,38 +684,297 @@ class PubSub:
...
+    def debug_object(self, key) -> Pipeline: ...
+    def echo(self, value) -> Pipeline: ...
+    def flushall(self) -> Pipeline: ...
+    def flushdb(self) -> Pipeline: ...
+    def info(self, section: Optional[_Key] = ...) -> Pipeline: ...
+    def lastsave(self) -> Pipeline: ...
+    def object(self, infotype, key) -> Pipeline: ...
+    def ping(self) -> Pipeline: ...
+    def save(self) -> Pipeline: ...
+    def sentinel(self, *args) -> None: ...
```
Comment (chdsbd) [inline stubs/redis/redis/client.pyi:766]: this deprecated method doesn't do or return anything: https://github.com/andymccurdy/redis-py/blob/1a41cfd95a53a95b078084d8627be6b6fba3bb71/redis/client.py#L1448-L1451

File: stubs/redis/redis/client.pyi (hunk 2 above)
Comment (srittau) [reply stubs/redis/redis/client.pyi:723]: This probably was leftover from an old redis-py version.

File: stubs/redis/redis/client.pyi (hunk 10 above)
Comment (srittau) [reply stubs/redis/redis/client.pyi:831]: Personally, I try to use the order the implementation uses. But it is your choice.

**Review by srittau — CHANGES_REQUESTED** [review]
Thank you for this PR and sorry for the long delay. Only one issue below.

**Comment by chdsbd** [comment]
I updated `Pipeline` to inherit from Redis and added `# type: ignore [override]` to every method that returns `Pipeline`. Not all of these ignore statements are necessary because some methods in parent class are missing return types.

The ordering of methods should match the `redis-py` implementation. The `Pipeline` defines a few methods but inherits most methods from the `Redis` client.

**Review by srittau — APPROVED** [review]

**Comment by Akuli** [comment]
I think `# type: ignore[override]` is a non-standard mypy thing? This PR however uses `# type: ignore [override]`, with a space. Maybe type checkers understand that to be same as just `# type: ignore`?

**Pull request merged**